File: src/ts/src/unity_ads/internal/UnityInterstitialAd.ts
import {
    AdObserver,
    FullScreenAdResult,
    IFullScreenAd,
} from "../../ads";
import { IAsyncHelper } from "../../ads/internal";
import {
    ObserverManager,
    Utils,
} from "../../core";
import { UnityAds } from "./UnityAds";

export class UnityInterstitialAd extends ObserverManager<AdObserver> implements IFullScreenAd {
    private readonly _displayer: IAsyncHelper<FullScreenAdResult>;
    private readonly _plugin: UnityAds;
    private readonly _adId: string;

    public constructor(
        displayer: IAsyncHelper<FullScreenAdResult>, plugin: UnityAds, adId: string) {
        super();
        this._displayer = displayer;
        this._plugin = plugin;
        this._adId = adId;
    }

    public destroy(): void {
        this._plugin.destroyAd(this._adId);
    }

    public get isLoaded(): boolean {
        return this._plugin.hasRewardedAd(this._adId);
    }

    public async load(): Promise<boolean> {
        Utils.noAwait(async () => { //
            await this._plugin.loadRewardedAd(this._adId);
        });
        return this.isLoaded;
    }

    public show(): Promise<FullScreenAdResult> {
        return this._displayer.process(
            () => this._plugin.showRewardedAd(this._adId),
            result => {
                // OK.
            });
    }

    public onLoaded(): void {
        this.dispatchEvent(observer => observer.onLoaded && observer.onLoaded());
    }

    public onFailedToShow(message: string): void {
        if (this._displayer.isProcessing) {
            this._displayer.resolve(FullScreenAdResult.Failed);
        } else {
            // Assert.
        }
    }

    public onClosed(): void {
        if (this._displayer.isProcessing) {
            this._displayer.resolve(FullScreenAdResult.Completed);
        } else {
            // Assert.
        }
    }
}

File: src/unity/Runtime/UnityAds/Internal/UnityInterstitialAd.cs
using System.Threading.Tasks;

using UnityEngine.Assertions;

namespace EE.Internal {
    internal class UnityInterstitialAd : ObserverManager<AdObserver>, IFullScreenAd {
        private readonly IAsyncHelper<AdResult> _displayer;
        private readonly UnityAds _plugin;
        private readonly string _adId;

        public UnityInterstitialAd(
            IAsyncHelper<AdResult> displayer, UnityAds plugin, string adId) {
            _displayer = displayer;
            _plugin = plugin;
            _adId = adId;
        }

        public void Destroy() {
            _plugin.DestroyFullScreenAd(_adId);
        }

        public bool IsLoaded => _plugin.HasRewardedAd(_adId);

        public Task<bool> Load() {
            Utils.NoAwait(async () => { //
                await _plugin.LoadRewardedAd(_adId);
            });
            return Task.FromResult(IsLoaded);
        }

        public Task<AdResult> Show() {
            return _displayer.Process(
                () => _plugin.ShowRewardedAd(_adId),
                result => {
                    // OK.
                });
        }

        internal void OnLoaded() {
            DispatchEvent(observer => observer?.OnLoaded());
        }

        internal void OnFailedToShow(string message) {
            if (_displayer.IsProcessing) {
                _displayer.Resolve(AdResult.Failed);
            } else {
                Assert.IsTrue(false);
            }
        }

        internal void OnClosed() {
            if (_displayer.IsProcessing) {
                _displayer.Resolve(AdResult.Completed);
            } else {
                Assert.IsTrue(false);
            }
        }
    }
}